swap/cheque.go

```diff
@@ -0,0 +1,125 @@
+// Copyright 2019 The Swarm Authors
+// This file is part of the Swarm library.
+//
+// The Swarm library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The Swarm library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the Swarm library. If not, see <http://www.gnu.org/licenses/>.
+
+package swap
+
+import (
+	"bytes"
+	"crypto/ecdsa"
+	"encoding/binary"
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+// MarshallBinary encodes the cheque in the format used in the signing procedure
+func (cheque *Cheque) MarshallBinary() ([]byte, error) {
```

Comment:
We didn't need `UnmarshalBinary` because this is only used for signature generation and there is never any need to decode from that encoded form. It also would be impossible to properly implement it as not every field from `Cheque` struct is part of the encoded form.

Implementing the `BinaryMarshaler` interface properly also creates problems. Since `MarshallBinary` is also used by `Store.put` when saving the cheque it wouldn't store the complete cheque. One solution might be to put the function on `ChequeParams` instead (and move `honey` to `Cheque`, since this also is not part of the encoding), however in that case we would have to manually implement `BinaryUnmarshaler` which otherwise seems to be automatic (?).

Alternatively we could revert this change, not implement `BinaryMarshaler` and call it `encode` or `encodeForSignature` again since this is a different kind of encoding (the format the smart contract expects to be used for signing vs the serialisation format for persistence).

Reply:
I am in favor of the alternative, i.e. renaming this function.

@holisticode thoughts?

Reply:
Yes, I agree, the actual semantics of the `BinaryMarshaler` can not be met by the functionality we need, thus we shouldn't then probably even implement it. We don't need the full `Cheque` object in binary form, only a subset of it. And we only need it one-way.

Reply:
cool. rename then

Reply:
Fixed. Went with `encodeForSignature` to make it clear what this encoding is for.

Reply:
thank you @ralph-pichler!